src/lib/factories.ts

```diff
@@ -71,63 +71,82 @@ function resolveFactory(type: FactoryTypes, definition: CustomElementDefinition
-				throw new Error(`Could not resolve '${mid}' to ${errorStrings[type]} factory function`);
+	else {
+		return new Promise((resolve, reject) => {
+			if (instance) {
+				resolveMid<Instance>(instance)
+					.then((defaultExport) => {
+						if (!defaultExport || typeof defaultExport !== 'object') {
+							reject(new Error(`Could not resolve '${instance}' to ${errorStrings[type]} instance`));
+						}
+						else {
+							resolve(() => defaultExport);
+						}
+					})
+					.catch(reject);
 			}
-
-			const factory: Factory = defaultExport;
-			return factory;
-		}
-
-		// istanbul ignore else Action factories are expected to guard against definitions with neither
-		// factory or instance properties.
-		if (instance) {
-			if (!defaultExport || typeof defaultExport !== 'object') {
-				throw new Error(`Could not resolve '${mid}' to ${errorStrings[type]} instance`);
+			else {
+				resolveMid<Factory>(factory)
+					.then((defaultExport) => {
+						if (typeof defaultExport !== 'function') {
+							reject(new Error(`Could not resolve '${factory}' to ${errorStrings[type]} factory function`));
+						}
+						else {
+							resolve(defaultExport);
+						}
+					})
+					.catch(reject);
 			}
-
-			const instance: Instance = defaultExport;
-			return () => instance;
-		}
-	});
+		});
+	}
 }
 
-export function makeActionFactory(definition: ActionDefinition, resolveMid: ResolveMid): ActionFactory {
+export function makeActionFactory(definition: ActionDefinition, resolveMid: ResolveMid, registry: ReadOnlyRegistry): ActionFactory {
 	if (!('factory' in definition || 'instance' in definition)) {
 		throw new TypeError('Action definitions must specify either the factory or instance option');
 	}
-	if ('instance' in definition && 'stateFrom' in definition) {
-		throw new TypeError('Cannot specify stateFrom option when action definition points directly at an instance');
+	if ('instance' in definition) {
+		if ('state' in definition) {
+			throw new TypeError('Cannot specify state option when action definition points directly at an instance');
+		}
+		if ('stateFrom' in definition) {
+			throw new TypeError('Cannot specify stateFrom option when action definition points directly at an instance');
+		}
 	}
 
-	return (registry: CombinedRegistry, defaultActionStore: StoreLike) => {
+	const { id, state: initialState } = definition;
+	return ({ registryProvider, stateFrom: defaultActionStore }) => {
 		return Promise.all<any>([
 			resolveFactory('action', definition, resolveMid),
 			resolveStore(registry, definition)
 		]).then(([_factory, _store]) => {
 			const factory = <ActionFactory> _factory;
 			const store = <StoreLike> _store || defaultActionStore;
 
-			return factory(registry, store);
+			const options = { registryProvider, stateFrom: store };
+
+			if (store && initialState) {
+				return store.add(initialState, { id })
+					// Ignore error, assume store already contains state for this widget.
+					.catch(() => undefined)
```

Comment:
Is there a valid reason for returning `undefined`, or is it really just a noop (`() => {}`)?